qutebrowser/mainwindow/tabbedbrowser.py

```diff
@@ -697,6 +697,9 @@ def on_current_changed(self, idx):
             log.webview.debug("on_current_changed got called with invalid "
                               "index {}".format(idx))
             return
+        if (self._now_focused is not None and
+                self._now_focused.tab_id is tab.tab_id):
+            return
```

Comment:
I haven't tested this yet, but are you sure it's safe to return here? We're missing a lot of logic below (such as setting the tab focus and leaving modes). I think we just want to avoid the last-focused-tab change if the tab is the same, right?

Reply:
This has the "side effect" of staying in insert mode when moving current tab, which I'd consider expected behavior. I'm not sure about the meaning of `setFocus()` on already focused tab. `new_instance_open_target` behaviour isn't changed.

Reply:
Hmm, sounds like a good change (to me), could you add a test for the insert mode staying so it sticks?

I'm still getting a few focusing issues, which is why I'm paranoid about the focus change, but it probably dosen't matter (the issues are probably coming from elsewhere)

Reply:
Looking at the test results, it looks like some of this stuff is needed, these tests look like they are failing:

```
FAILED tests/end2end/features/test_keyinput_bdd.py::test_mode_on_change_restore
FAILED tests/end2end/features/test_open_bdd.py::test_opening_in_a_new_tab_related
FAILED tests/end2end/features/test_tabs_bdd.py::test_undo_a_tab_closed_after_rearranging_tabs
```

Could you take a look at that? I think the 'tab_insert_idx_{left,right}' stuff might be needed, and maybe the signal too.

Reply:
- Test `mode_on_change_restore` is passing for me.
- I think test `undo_a_tab_closed_after_rearranging_tabs` is wrong and I actually fixed the behavior – the step `:tab-focus 2` doesn't do anything on master. I've updated the test and also added a new one for `:tab-focus last` after closing a lower number tab, also failing on master.
- Test `opening_in_a_new_tab_related` failing indeed is a regression and seems to depend on only `self._tab_insert_idx_{left,right}`. Should I just move those up? I didn't find problems with skipping `current_tab_changed.emit()`.

Reply:
>  is wrong and I actually fixed the behavior

Oh, I see, before, the `tab-focus` would cause the tab to stay on tab 2, instead of switching to the last like documented, nice :).

>  Should I just move those up?

I think this should be safe, I'm going to play with this a little to try to be more sure though.

>Test mode_on_change_restore is passing for me.

This is passing for me too, but it's definitely failing on CI consistently (https://travis-ci.org/qutebrowser/qutebrowser/jobs/509265658). I'm not sure how that could be happening. My best guess for that is something in the signal is not getting triggered for some reason anymore. I don't know enough about the CI system to comment on why there's a difference between the environments though, maybe it's a race condition.